src/uu/chcpu/src/chcpu.rs

```diff
@@ -0,0 +1,379 @@
+
+// Takes in a human-readable list of CPUs, and returns a list of indices parsed from that list
+// These can come in the form of a plain range like `X-Y`, or a comma-separated ranges and indices ie. `1,3-4,7-8,10`
+// Kernel docs with examples: https://www.kernel.org/doc/html/latest/admin-guide/cputopology.html
+fn parse_cpu_list(list: &str) -> Vec<usize> {
+    let mut out: Vec<usize> = vec![];
+
+    if list.is_empty() {
+        return out;
+    }
+
+    for part in list.trim().split(",") {
+        if part.contains("-") {
+            let bounds: Vec<_> = part.split("-").flat_map(|x| x.parse::<usize>()).collect();
+            assert_eq!(bounds.len(), 2);
+            for idx in bounds[0]..bounds[1] + 1 {
+                out.push(idx)
+            }
+        } else {
+            let idx = part.parse::<usize>().expect("Invalid CPU index value");
+            out.push(idx);
+        }
+    }
+    out
+}
+
+#[test]
+fn test_parse_cpu_list() {
+    assert_eq!(parse_cpu_list(""), Vec::<usize>::new());
+    assert_eq!(parse_cpu_list("1-3"), Vec::<usize>::from([1, 2, 3]));
+    assert_eq!(parse_cpu_list("1,2,3"), Vec::<usize>::from([1, 2, 3]));
+    assert_eq!(
+        parse_cpu_list("1,3-6,8"),
+        Vec::<usize>::from([1, 3, 4, 5, 6, 8])
+    );
+    assert_eq!(
+        parse_cpu_list("1-2,3-5,7"),
+        Vec::<usize>::from([1, 2, 3, 4, 5, 7])
+    );
+}
+
+#[derive(Debug)]
+struct Cpu(usize);
+
+impl Cpu {
+    fn get_path(&self) -> PathBuf {
+        PathBuf::from(format!("/sys/devices/system/cpu/cpu{}", self.0))
+    }
+
+    fn ensure_exists(&self) -> bool {
+        if !self.get_path().exists() {
+            println!("CPU {} does not exist", self.0);
+            return false;
+        };
+        true
+    }
+
+    // CPUs which are not hot-pluggable will not have the `/online` file in their directory
+    fn is_hotpluggable(&self) -> bool {
+        self.get_path().join("online").exists()
+    }
+
+    fn is_online(&self) -> bool {
+        if let Ok(state) = fs::read_to_string(self.get_path().join("online")) {
+            match state.trim() {
+                "0" => return false,
+                "1" => return true,
+                other => panic!("Unrecognized CPU online state: {}", other),
+            }
+        };
+
+        // Just in case the caller forgot to check `is_hotpluggable` first,
+        // instead of panicing that the file doesn't exist, return true
+        // This is because a non-hotpluggable CPU is assumed to be always online
+        true
+    }
+
+    fn enable(&self) {
+        if !self.is_hotpluggable() {
+            println!("CPU {} is not hot-pluggable", self.0);
+            return;
+        }
+
+        if self.is_online() {
+            println!("CPU {} is already enabled", self.0);
+            return;
+        }
+
+        let result =
+            File::create(self.get_path().join("online")).and_then(|mut f| f.write_all(b"1"));
```

Comment:
looks like you are duplicating the code a bit in these 3 functions.
what about creating something like ?

fn write_cpu_state(path: PathBuf, value: &[u8], action: &str, cpu_id: u32) {
    let result = File::create(path).and_then(|mut f| f.write_all(value));
    match result {
        Ok(_) => println!("CPU {} {}", cpu_id, action),
        Err(e) => println!("CPU {} {} failed: {}", cpu_id, action, e.kind()),
    }
}